### PR TITLE
clusterresolver: deflake eds_impl tests

### PR DESCRIPTION
**Pull request by easwars**
The current tests in `xds/internal/balancer/clusterresolver/eds_impl_test.go`, which verify EDS implementation in the `clusterresolver` LB policy have been extremely flaky of late. This PR makes the following changes:
- Add a `roundrobin` package to `internal/testutils` which provides helper functions to verify that RPCs are roundrobin-ed across the given set of addresses. e2e tests which verify RPCs on the `TestService` can make use of this package.
  - Change existing tests in `test/balancer_switching_test.go` and `test/roundrobin_test.go` to use the above package
- Rewrite existing tests in `eds_impl_test.go` as e2e tests, which do the following:
  - Setup an xDS management server
  - Setup backends exposing the `TestService`
  - Create an `xdsclient` to talk to the above management server
  - Configure a manual resolver which configures the `cluster_resolver_experimental` LB policy on the client channel
  - Configure resources on the xDS management server and verify expected behavior on the client channel
- Get rid of tests which were verifying drops and circuit_breaking functionality. These are actually provided by the `clusterimpl` LB policy and we have tests there which verify this functionality.

I have verified that these tests don't flake in google3 for over 100K runs.

Fixes https://github.com/grpc/grpc-go/issues/4550, https://github.com/grpc/grpc-go/issues/4697, https://github.com/grpc/grpc-go/issues/5465

RELEASE NOTES: none

### CODE REVIEW

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+	resources = clientResources(nodeID, edsServiceName, nil)
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := waitForTransientFailure(ctx, t, testClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func waitForTransientFailure(ctx context.Context, t *testing.T, client testgrpc.TestServiceClient) error {
```
Comment (zasweq) [inline xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:540]: Nit: are we really waiting for transient failure here specifically? I feel like the ClientConn can go TRANSIENT_FAILURE in an orthogonal situation than this specific error defined in the priority package. Can you please change name/add top level comment for function.

File: internal/testutils/xds/e2e/setup_management_server.go [hunk 2]
```diff
@@ -89,12 +89,16 @@ func SetupManagementServer(t *testing.T, opts *ManagementServerOptions) (*Manage
 		server.Stop()
 		t.Fatalf("Failed to create bootstrap file: %v", err)
 	}
-	resolverBuilder := internal.NewXDSResolverWithConfigForTesting.(func([]byte) (resolver.Builder, error))
-	resolver, err := resolverBuilder(bootstrapContents)
-	if err != nil {
-		server.Stop()
-		t.Fatalf("Failed to create xDS resolver for testing: %v", err)
+
+	var r resolver.Builder
```
Comment (zasweq) [inline internal/testutils/xds/e2e/setup_management_server.go:93]: Nit: would prefer rb.

**Review by zasweq — CHANGES_REQUESTED** [review]
Some comments. Thanks for being proactive about test maintence/test flakes.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+			t.Logf("EmptyCall() returned code: %v, want: %v", code, codes.Unavailable)
+			continue
+		}
+		if !strings.Contains(err.Error(), priority.ErrAllPrioritiesRemoved.Error()) {
+			t.Logf("EmptyCall() = %v, want %v", err, priority.ErrAllPrioritiesRemoved)
+			continue
+		}
+		return nil
+	}
+	return errors.New("Timeout when waiting for RPCs to fail with UNAVAILABLE status")
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go]: Done.

File: xds/internal/balancer/clusterresolver/eds_impl_test.go [hunk 4]
```diff
@@ -1,569 +0,0 @@
...
-		t.Fatalf("RemoveSubConn, want %v, got %v", sc2, scToRemove)
-	}
-	edsb.UpdateSubConnState(scToRemove, balancer.SubConnState{ConnectivityState: connectivity.Shutdown})
-
-	// Test pick with only the third subconn.
-	if err := testRoundRobinPickerFromCh(cc.NewPickerCh, []balancer.SubConn{sc3}); err != nil {
-		t.Fatal(err)
-	}
-
-	// The same locality, different drop rate, dropping 50%.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/eds_impl_test.go:167]: We possibly can, but it is being tested in the cluster_impl LB policy which is the one which implements drops. Historically, all this functionality was in a single LB policy, and therefore all tests lived here. But over time the EDS LB policy has been split up into multiple ones, but the tests havent been cleaned up here.

**Review by easwars — COMMENTED** [review]
Thanks for the detailed review.

File: internal/testutils/roundrobin/roundrobin.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	// There is a small possibility that RPCs are still getting routed to a
+	// backend which was recently removed from the list of addresses passed to
+	// the round_robin policy. This is usually a transitory state and will
+	// eventually fix itself, and RPCs will start getting routed to only
+	// backends that we care about. We work around this situation by using two
+	// loops. The inner loop contains the meat of the calculations. If we ever
+	// see an RPCs getting routed to a backend that we dont expect it to get
+	// routed to, we break from the inner loop thereby resetting all the
+	// internal state and start afresh.
+OuterLoop:
```
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:166]: Why does this even need an outer loop? Why don't you just return an error in line 173, I think it is logically equivalent.

File: internal/testutils/roundrobin/roundrobin.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+				gotRatio[addr] = count / totalCount
+			}
+			if equalApproximate(gotRatio, wantRatio) {
+				return nil
+			}
+			logger.Infof("non-weighted-roundrobin, gotRatio: %v, wantRatio: %v", gotRatio, wantRatio)
+		}
+		<-time.After(time.Millisecond)
+	}
+	return fmt.Errorf("Timeout when waiting for roundrobin distribution of RPCs across addresses: %v", addrs)
```
Comment (zasweq) [inline internal/testutils/roundrobin/roundrobin.go:198]: Non capital first letter in error string.

File: internal/testutils/roundrobin/roundrobin.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:166]: Inner loop can be something like: tryToGetDistribution or something along those lines.

File: internal/testutils/roundrobin/roundrobin.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+// Returns a non-nil error if context deadline expires before RPCs start to get
+// roundrobin-ed across the given backends.
+func CheckWeightedRoundRobinRPCs(ctx context.Context, client testgrpc.TestServiceClient, addrs []resolver.Address) error {
+	if err := waitForTrafficToReachBackends(ctx, client, addrs); err != nil {
+		return err
+	}
+
+	// At this point, RPCs are getting successfully executed at the backends
+	// that we care about. To take the randomness of the WRR into account, we
+	// look for approximate distribution instead of exact.
```
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:147]: This feels really weird to me. I get that we're encapsulating psuedo random numbers in an e2e test with the randomness affecting RPC's, but continuously retrying until it falls within a distribution, especially with how many times this is called just feels off to me. I get that mathmatically perhaps "eventually factors out the randomness and gets it to be within our 5% range." But still "eventually factors out the randomness" is too strong an assertion to make here. I don't know if there is any better way to test this though since pseudo random + e2e. What are your thoughts/suggestions on this @dfawley?

File: internal/testutils/roundrobin/roundrobin.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+				}
+				results[peer.Addr.String()]++
+			}
+			totalCount += float64(len(addrs))
+
+			gotRatio := make(map[string]float64)
+			for addr, count := range results {
+				gotRatio[addr] = count / totalCount
+			}
+			if equalApproximate(gotRatio, wantRatio) {
```
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:191]: Hmmm ok thanks for the explanation. Worth noting that this is the logic which also "factors out the randomness" according to your other comments.

File: internal/testutils/roundrobin/roundrobin.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:166]: If so on either update comment please. Also perhaps move the comment block into more specific places in this block of code. It is hard to follow what is happening here.

**Review by zasweq — CHANGES_REQUESTED** [review]
Some comments.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func backendAddressesAndPorts(t *testing.T, servers []*stubserver.StubServer) ([]resolver.Address, []uint32) {
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:73]: Done, thanks.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+	servers := make([]*stubserver.StubServer, numBackends)
+	for i := 0; i < numBackends; i++ {
+		servers[i] = &stubserver.StubServer{
+			EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) { return &testpb.Empty{}, nil },
+		}
+		servers[i].StartServer()
+	}
+
+	return servers, func() {
+		for i := 0; i < numBackends; i++ {
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:105]: I think I just forced myself to use `numBackends` because I have it. Changed it.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+// service name and localities.
+func clientEndpointsResource(nodeID, edsServiceName string, localities []localityInfo) e2e.UpdateOptions {
+	return e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{endpointResource(edsServiceName, localities)},
+		SkipValidation: true,
+	}
+}
+
+// TestEDS_OneLocality tests the cluster_resolver LB policy using an EDS
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:166]: I don't understand. Do you want something changed here?

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+	managementServer, nodeID, bootstrapContents, _, cleanup1 := e2e.SetupManagementServer(t, nil)
+	defer cleanup1()
+
+	// Start backend servers which provide an implementation of the TestService.
+	servers, cleanup2 := startTestServiceBackends(t, 4)
+	defer cleanup2()
+	addrs, ports := backendAddressesAndPorts(t, servers)
+
+	// Create xDS resources for consumption by the test. We start off with a
+	// two localities, and single backend in each of them.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:283]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+		{name: localityName1, weight: 1, ports: ports[:1]},
+		{name: localityName2, weight: 1, ports: ports[1:2]},
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create an xDS client for use by the clusterresolver LB policy.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:294]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+						"edsServiceName": "%s"
+					}],
+					"xdsLbPolicy":[{"round_robin":{}}]
+				}
+			}]
+		}`, clusterName, edsServiceName)
+	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, client))
+
+	// Create a ClientConn and make a successful RPC.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:319]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+	testClient := testpb.NewTestServiceClient(cc)
+	if err := rrutil.CheckWeightedRoundRobinRPCs(ctx, testClient, addrs[0:2]); err != nil {
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:326]: I have changed all inline comments to state "weighted roundrobin". I also added a top-level comment talking about weighted roundrobin and the randomness associated with it. I feel we have gone above and beyond wrt documenting these tests :)

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 16]
```diff
@@ -0,0 +1,558 @@
...
+	// Add another locality with a single backend, and ensure roundrobin.
+	resources = clientResources(nodeID, edsServiceName, []localityInfo{
+		{name: localityName1, weight: 1, ports: ports[:1]},
+		{name: localityName2, weight: 1, ports: ports[1:2]},
+		{name: localityName3, weight: 1, ports: ports[2:3]},
+	})
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := rrutil.CheckWeightedRoundRobinRPCs(ctx, testClient, addrs[0:3]); err != nil {
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:339]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 17]
```diff
@@ -0,0 +1,558 @@
...
+
+	// Remove the first locality, and ensure roundrobin.
+	resources = clientResources(nodeID, edsServiceName, []localityInfo{
+		{name: localityName2, weight: 1, ports: ports[1:2]},
+		{name: localityName3, weight: 1, ports: ports[2:3]},
+	})
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := rrutil.CheckWeightedRoundRobinRPCs(ctx, testClient, addrs[1:3]); err != nil {
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:351]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 18]
```diff
@@ -0,0 +1,568 @@
...
+			v3corepb.HealthStatus_DEGRADED,
+		}},
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create an xDS client for use by the clusterresolver LB policy.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:427]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 19]
```diff
@@ -0,0 +1,558 @@
...
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, client))
+
+	// Create a ClientConn and make a successful RPC.
+	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+	testClient := testpb.NewTestServiceClient(cc)
+	if err := rrutil.CheckWeightedRoundRobinRPCs(ctx, testClient, append(addrs[0:2], addrs[6:8]...)); err != nil {
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:459]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 20]
```diff
@@ -0,0 +1,568 @@
...
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:459]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 21]
```diff
@@ -0,0 +1,568 @@
...
+		t.Fatal(err)
+	}
+}
+
+// TestEDS_EmptyUpdate tests the cluster_resolver LB policy using an EDS
+// resource with no localities and verifies that RPCs fail with the expected
+// error.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:466]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 22]
```diff
@@ -0,0 +1,568 @@
...
+
+	// Start backend servers which provide an implementation of the TestService.
+	servers, cleanup2 := startTestServiceBackends(t, 4)
+	defer cleanup2()
+	addrs, ports := backendAddressesAndPorts(t, servers)
+
+	const cacheTimeout = 100 * time.Microsecond
+	oldCacheTimeout := balancergroup.DefaultSubBalancerCloseTimeout
+	balancergroup.DefaultSubBalancerCloseTimeout = cacheTimeout
+	defer func() { balancergroup.DefaultSubBalancerCloseTimeout = oldCacheTimeout }()
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:480]: Got rid of the const. But didn't change it to the version you have here. Just my preference I guess.

Yes, we love the metric system :)

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 23]
```diff
@@ -0,0 +1,568 @@
...
+	// Create xDS resources for consumption by the test. The first update is an
+	// empty update. This should put the channel in TRANSIENT_FAILURE.
+	resources := clientEndpointsResource(nodeID, edsServiceName, nil)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create an xDS client for use by the clusterresolver LB policy.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:491]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 24]
```diff
@@ -0,0 +1,568 @@
...
+						"edsServiceName": "%s"
+					}],
+					"xdsLbPolicy":[{"round_robin":{}}]
+				}
+			}]
+		}`, clusterName, edsServiceName)
+	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, client))
+
+	// Create a ClientConn and ensure TRANSIENT_FAILURE.
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:516]: Done.

**Review by easwars — COMMENTED** [review]
Thanks for the review again.

**Comment by easwars** [comment]
Also, if you have more comments, it would be great if you could mark the ones which have been handled satisfactorily as "resolved". That way, we can track just the open comments. Thanks.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 25]
```diff
@@ -70,6 +70,9 @@ func Test(t *testing.T) {
 	grpctest.RunSubTests(t, s{})
 }
 
+// "backendAddressesAndPorts extracts the address and port of each of the
```
Comment (zasweq) [inline xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:73]: Remove this " :).

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 26]
```diff
@@ -367,10 +380,11 @@ func (s) TestEDS_MultipleLocalities(t *testing.T) {
...
 
-	// Change the weight of one locality and ensure weighted roundrobin.  Since
-	// RPCs are weighted roundrobined across localities, and locality2 has two
-	// times the weight of locality1, locality2's backend will receive four
-	// times the traffic of locality1's backend.
+	// Change the weight of locality2 and ensure weighted roundrobin.  Since
+	// locality2 has twice the weight of locality3, it will be picked twice as
+	// frequently as locality3 for RPCs. And since locality2 has half the number
+	// of backends as locality3, its backends will receive four times the
+	// traffic of locality3's backend.
```
Comment (zasweq) [inline xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:387]: Sorry, I honestly don't care that much, but this is still logically incorrect. Four times the traffic of *either* one of locality3's backend. Or switch to two times the traffic of locality3.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 27]
```diff
@@ -533,7 +551,8 @@ func (s) TestEDS_EmptyUpdate(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Push another empty update and ensure TRANSIENT_FAILURE.
+	// Push another empty update and ensure that RPCs fails with "all priorities
```
Comment (zasweq) [inline xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:554]: Nit: s "fails" to "fail".

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go [hunk 28]
```diff
@@ -340,7 +352,8 @@ func (s) TestEDS_MultipleLocalities(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Remove the first locality, and ensure roundrobin.
+	// Remove the first locality, and ensure RPCs are being weighted
+	// roundrobined across the last two backends.
```
Comment (zasweq) [inline xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:356]: Nit: switch "last two backends" to "two backends remaining" or something.

File: internal/testutils/roundrobin/roundrobin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:147]: I don't know. It feels weird to wrap something like randomness with something like a statistical wrapper around the randomness with, even if it has whatever percentage chance of the randomness falling within a distribution. I chatted with Doug about this, and he mentioned that there's no way around this with regards to randomness. Thus, I am fine with it, even though it feels icky to me.

File: internal/testutils/roundrobin/roundrobin.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply internal/testutils/roundrobin/roundrobin.go:175]: I get the point of breaking out of the inter loop and starting a fresh. But the jump to the OuterLoop on a context error (the only place you break to that loop returns fmt.Errorf(...). I'm saying it is equivalent to return that in line 182, and not even have this OuterLoop label thing. Thanks for the comment. It makes me realize I was viewing this wait wrongly. This just gives it more and more time to hit the expected distribution, so I fine with it now, it makes more sense that way, it's like a mathematical limit/asymptote.

**Review by zasweq — APPROVED** [review]
LGTM outside minor nits.

File: internal/testutils/roundrobin/roundrobin.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (easwars) [reply internal/testutils/roundrobin/roundrobin.go:175]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go (hunk 25 above)
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:73]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go (hunk 28 above)
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:356]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go (hunk 26 above)
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:387]: Done.

File: xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go (hunk 27 above)
Comment (easwars) [reply xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go:554]: Done.
